File: backend/internal/domain/domainrule.go
package domain

import "time"

type RuleType string

const (
	RuleTypeAllow RuleType = "allow"
	RuleTypeDeny  RuleType = "deny"
)

type RuleKind string

const (
	RuleKindExact RuleKind = "exact"
	RuleKindRegex RuleKind = "regex"
)

type DomainRule struct {
	Domain    string
	Unicode   string
	Type      RuleType
	Kind      RuleKind
	Comment   *string
	Groups    []int
	Enabled   bool
	Id        int
	CreatedAt time.Time
	UpdatedAt time.Time
}

type DomainRuleSet struct {
	Rules []DomainRule
	Took  time.Duration
}

// Queries/Commands your service & cluster expose
type ListDomainRulesQuery struct {
	Type   *RuleType
	Kind   *RuleKind
	Domain *string
}

type AddDomainRulesCommand struct {
	Type    RuleType
	Kind    RuleKind
	Domains []string
	Comment *string
	Groups  []int
	Enabled *bool
}

type RemoveDomainRuleCommand struct {
	Type   RuleType
	Kind   RuleKind
	Domain string
}

// Add Domain
type DomainProcessedItem struct {
	Item string
}

type DomainProcessedError struct {
	Item  string
	Error string
}

type DomainProcessed struct {
	Success []DomainProcessedItem
	Errors  []DomainProcessedError
}

type AddDomainRulesResult struct {
	Rules     []DomainRule
	Processed DomainProcessed
	Took      time.Duration
}

type SyncDomainRuleCommand struct {
	Type    RuleType
	Kind    RuleKind
	Domain  string
	Comment *string
}

type SyncDomainRuleNodeResult struct {
	AlreadyPresent bool
	Added          bool
}

type RegexMatch struct {
	ID      int
	Pattern string
	Kind    string // "deny" | "allow"
	Enabled bool
}

type RegexTestResult struct {
	Domain  string
	Matches []RegexMatch
}

File: frontend/src/types/domainrule.ts
import type { PiholeNodeRef } from './pihole';

export type RuleType = 'allow' | 'deny';
export type RuleKind = 'exact' | 'regex';

export type DomainRule = {
	id: number;
	domain: string;
	unicode: string;
	type: RuleType;
	kind: RuleKind;
	comment?: string | null;
	groups: number[];
	enabled: boolean;
	createdAt: string;
	updatedAt: string;
};

export type ListNodeResult = {
	node: PiholeNodeRef;
	rules: DomainRule[];
	tookMs: number;
	error?: string;
};

export type ListSummary = {
	totalNodes: number;
	okNodes: number;
	errorNodes: number;
	totalRules: number;
};

export type ListDomainRulesResponse = {
	summary: ListSummary;
	nodes: Record<string, ListNodeResult>;
};

export type AddDomainRuleResponse = {
	nodes: Record<string, {
		node: PiholeNodeRef;
		result: {
			domains: DomainRule[];
			processed: {
				success: { item: string }[];
				errors: { item: string; error: string }[];
			};
			tookMs: number;
		};
		error?: string;
	}>;
};

export type RemoveDomainRuleResponse = {
	summary: {
		total: number;
		removed: number;
		failed: number;
		errors: number;
	};
	nodes: Record<string, {
		node: PiholeNodeRef;
		removed: boolean;
		error?: string;
	}>;
};

export type ConsolidatedRule = {
	key: string;
	domain: string;
	type: RuleType;
	kind: RuleKind;
	enabled: boolean;
	comment?: string | null;
	nodeIds: number[];
	totalNodes: number;
};
